### DISCUSSION TITLE
How can multi-action buttons propagate state?

### DISCUSSION THREAD

**Discussion by buffos** [elgatosf/elgato#143] [Help] [answered]
How can buttons within a multi-action that do not belong to the same plugin pass values between them?

For example, I have a button that makes an HTTP request and gets me a result. Then I use that result as an input to another button that makes another action.

Of course, if I write all the buttons myself. They can communicate. But that means that I cannot use plugins written by someone else in a multi-action.

It would be nice if there were a cache for a multi-action doing something like

```
if (ev.payload.isInMultiAction) {
  value = ev.action.readFromCache()
 // do operations
}
// in another part of the code
if (ev.payload.isInMultiAction) {
   ev.action.writeToCache(obj) // to be used by the next action.
}
```

Is this possible in any way as of now?

**Comment by ChekTek** [comment]
Hi @buffos, 

Very cool idea! This is exact concept is not available at the moment, but plugins can communicate with one another via [deep-linking](https://docs.elgato.com/streamdeck/sdk/guides/deep-linking). i.e. when a plugin completes the http request you can then send that result to another plugin via deeplinking through the next plugins UUID. I do worry a bit, that the asynchronous nature of these events could result in some inconsistent results.  

Overall, I don't see something more advanced than this in the current SD roadmap, but I will relay the message to the team!

**Reply by buffos** [reply]
Hi @ChekTek 

Just a follow-up question. 

I receive the deep link, get the data, and update the globalSettings. That's clear.

Can I retrieve the globalSettings inside an Action? 

The documentation shows `await streamDeck.settings.getGlobalSettings` being used only at the plugin level and not at the Actions level.

**Reply by ChekTek** [reply, marked as answer]
Yup, you can import and utilize the `streamDeck` singleton anywhere throughout your plugin. Here is a quick snippet from our Twitch plugin that demonstrates using the logger from `streamDeck` within the `keyUp` of an action.
```ts
import streamDeck from "@elgato/streamdeck";

...

/**
 * Clears the chat of the specified Twitch channel.
 */
override async onKeyUp({ action, payload }: KeyUpEvent<ClearChatSettings>): Promise<void> {
	try {
		const { settings } = payload;
		const { accountId } = settings;

		await twitchService.clearChat(accountId);

		action.showOk();
	} catch (e) {
		action.showAlert();
		streamDeck.logger.error("Failed to clear chat", e);
	}
}

...
```

One thing worth noting, is that `getGlobalSettings` will also trigger a `didReceiveGlobalSettings` event to fire, if you have a listener set up in your plugin.

```ts
streamDeck.settings.onDidReceiveGlobalSettings(async (ev: DidReceiveGlobalSettingsEvent<GlobalSettings>) => {})
```
